### PR TITLE
[BUGFIX] Not working properly on self-hosted runners

### LINKED ISSUE #16: Support non-ephemeral, self-hosted runners
subsequent builds fail with:

```
Run webfactory/ssh-agent@v0.2.00s
##[error]Node run failed with exit code 1
Run webfactory/ssh-agent@v0.2.0
Adding GitHub.com keys to undefined/.ssh/known_hosts
Starting ssh-agent
bind: Address already in use
unix_listener: cannot bind to path: /tmp/ssh-auth.sock
##[error]Command failed: ssh-agent -a /tmp/ssh-auth.sock
bind: Address already in use
unix_listener: cannot bind to path: /tmp/ssh-auth.sock

##[error]Node run failed with exit code 1
```

This is on a local github-runners, which are not ephemeral. Cleaning up the /tmp/ssh-auth.sock manually resolves this problem temporarily.

### PR DESCRIPTION
**Pull request by thommyhh**
* remove fixed default auth socket path
* call `ssh-agent` with auth socket path, if not explicitly given,
  to let the system choose the path
* read SSH_AUTH_SOCKET and SSH_AGENT_PID from `ssh-agent` command output
  and export them as variables
* update `dist/index.js`
* update readme with section about exported variables

Fixes #16

### CODE REVIEW

**Comment by rbellamy** [comment]
Maybe I'm reading this incorrectly, but wouldn't this kill the agent for other runners on the same machine?

**Comment by rbellamy** [comment]
Adding the run ID seems like a wise choice to ensure the agent is using a unique `AUTH_SOCK`:

https://github.com/webfactory/ssh-agent/issues/16#issuecomment-580482355

**Comment by thommyhh** [comment]
@rbellamy As far as I know, `ssh-agent` picks a random unique file name for the auth socket, if none is provided. This is why I removed the default. By exporting the process ID `$SSH_AGENT_PID`, we can kill that one agent. No other agents are killed.

File: README.md [hunk 1]
```diff
@@ -57,6 +57,20 @@ The `ssh-agent` will load all of the keys and try each one in order when establi
 There's one **caveat**, though: SSH servers may abort the connection attempt after a number of mismatching keys have been presented. So if, for example, you have
 six different keys loaded into the `ssh-agent`, but the server aborts after five unknown keys, the last key (which might be the right one) will never even be tried.
 
+## Exported variables
+The action exports `SSH_AUTH_SOCK` and `SSH_AGENT_PID` through the Github Actions core module.
+The `$SSH_AUTH_SOCK` is used by several applications like git or rsync to connect to the SSH authentication agent.
+The `$SSH_AGENT_PID` can be used to kill the SSH agent, e.g. when using on a self-hosted runner:
```
Comment (mpdude) [inline README.md:63]: I'd shorten this to only say "... contains the process id of the agent running for this job". I think it should be the action's responsibility to kill the agent.

File: dist/index.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ try {
     const home = process.env['HOME'];
     const homeSsh = home + '/.ssh';
 
-    const privateKey = core.getInput('ssh-private-key').trim();
+    const privateKey = core.getInput('ssh-private-key');
```
Comment (mpdude) [inline dist/index.js:65]: Could you rebase this PR against `master` please? Then this change should go away.

File: dist/index.js [hunk 3]
```diff
@@ -71,14 +71,27 @@ try {
     }
 
     console.log(`Adding GitHub.com keys to ${homeSsh}/known_hosts`);
-    fs.mkdirSync(homeSsh, { recursive: true});
+    fs.mkdirSync(homeSsh, { recursive: true });
```
Comment (mpdude) [inline dist/index.js:74]: Same as before, should go away with a rebase.

File: dist/index.js [hunk 4]
```diff
@@ -71,14 +71,27 @@ try {
...
 
     console.log("Starting ssh-agent");
     const authSock = core.getInput('ssh-auth-sock');
-    child_process.execFileSync('ssh-agent', ['-a', authSock]);
-    core.exportVariable('SSH_AUTH_SOCK', authSock);
+    let sshAgentOutput = ''
+    if (authSock && authSock.length > 0) {
+        sshAgentOutput = child_process.execFileSync('ssh-agent', ['-a', authSock]);
+    } else {
+        sshAgentOutput = child_process.execFileSync('ssh-agent')
```
Comment (mpdude) [inline dist/index.js:84]: This is a slight BC break since now _not_ providing a value for the `ssh-auth-sock` input will bind it to a random port, instead of the fixed value we used previously. (Yes, I completely understand that this is part of the goal to be able to run multiple instances in parallel; just sayin').

Comment (mpdude) [reply dist/index.js:84]: Since we're still in the 0.x version ranges, IMO nothing we should worry about too much.

File: README.md (hunk 1 above)
Comment (thommyhh) [reply README.md]: I added a post action to kill the agent. Seems to work as expected.

File: dist/index.js (hunk 4 above)
Comment (thommyhh) [reply dist/index.js:84]: It binds to a random socket file, just for being precise. I don't think the value matters as long as it written to the `$SSH_AUTH_SOCK` variable.

**Comment by thommyhh** [comment]
@mpdude I updated the branch with the latest changes from master, but the changes did not disappear. I don't know, how to trigger GitHub to re-read the base branch (tried changing base).

I also added the post action to kill the agent automatically, as suggested. I wondered, if there is something like a post job script, to clean things up.

**Comment by zhigang1992** [comment]
I think you can merge master and then push?

**Comment by thommyhh** [comment]
@zhigang1992 
That's what I did, but GitHub did not update the base branch anyway.

**Comment by mpdude** [comment]
I have released 0.3.0 with this change. 

Thank you @thommyhh and everyone involved!

**Comment by mkrakowitzer** [comment]
Thanks @thommyhh @mpdude for resolving this!
